homeassistant/components/snapcast/coordinator.py

```diff
@@ -0,0 +1,75 @@
+"""Data update coordinator for Snapcast server."""
+
+from __future__ import annotations
+
+import logging
+
+from snapcast.control.server import Snapserver
+
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class SnapcastUpdateCoordinator(DataUpdateCoordinator[None]):
+    """Data update coordinator for pushed data from Snapcast server."""
+
+    def __init__(self, hass: HomeAssistant, host: str, port: int) -> None:
+        """Initialize coordinator."""
+        super().__init__(
+            hass,
+            logger=_LOGGER,
+            name=f"{host}:{port}",
+            update_interval=None,  # Disable update interval as server pushes
+        )
+
+        self._server = Snapserver(hass.loop, host, port, True)
+        self._available = False
+
+        self._server.set_on_update_callback(self._on_update)
+        self._server.set_new_client_callback(self._on_update)
+        self._server.set_on_connect_callback(self._on_connect)
+        self._server.set_on_disconnect_callback(self._on_disconnect)
+
+    def _on_update(self) -> None:
+        """Snapserver on_update callback."""
+        # Assume availability if an update is received.
+        self._available = True
+        self.async_set_updated_data(None)
+
+    def _on_connect(self) -> None:
+        """Snapserver on_connect callback."""
+        self._available = True
+        self.async_set_updated_data(None)
+
+    def _on_disconnect(self, ex):
+        """Snapsever on_disconnect callback."""
+        self._available = False
+        self.async_set_updated_data(None)
+
+    async def _async_setup(self) -> None:
+        """Perform async setup for the coordinator."""
+        # Start the server
+        await self._server.start()
+
+    async def _async_update_data(self) -> None:
+        """Empty update method since data is pushed."""
+
+    async def disconnect(self) -> None:
+        """Disconnect from the server."""
+        self._server.set_on_update_callback(None)
+        self._server.set_on_connect_callback(None)
+        self._server.set_on_disconnect_callback(None)
+        self._server.set_new_client_callback(None)
+        self._server.stop()
+
+    @property
+    def available(self) -> bool:
+        """Get availability of the server."""
+        return self._available
```

Comment:
I think we should replace all `self._available` with `self.last_update_success`

Reply:
Done.